File: app/src/main/java/com/example/sasham/testproject/messages/MessagesListingAdapter.java
package com.example.sasham.testproject.messages;

import android.support.v7.widget.RecyclerView;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.text.style.URLSpan;
import android.text.util.Linkify;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.example.sasham.testproject.R;
import com.example.sasham.testproject.model.Message;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.saket.bettermovementmethod.BetterLinkMovementMethod;

public class MessagesListingAdapter extends RecyclerView.Adapter<MessagesListingAdapter.ViewHolder> {


    private List<Message> messages;

    public MessagesListingAdapter(List<Message> messages) {
        this.messages = messages;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View root = LayoutInflater.from(parent.getContext()).inflate(R.layout.message_listing_item, parent, false);
        return new ViewHolder(root);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {

        Message currentMessage = messages.get(position);

        holder.userName.setText(currentMessage.getUserName());

        holder.messageText.setText(currentMessage.getMsgText());
        holder.messageText.setMovementMethod(BetterLinkMovementMethod.getInstance());
        Linkify.addLinks(holder.messageText,Linkify.WEB_URLS);
        BetterLinkMovementMethod.linkify(Linkify.WEB_URLS,holder.messageText)
                .setOnLinkClickListener(new BetterLinkMovementMethod.OnLinkClickListener() {
                    @Override
                    public boolean onClick(TextView textView, String url) {
                        Toast.makeText(holder.itemView.getContext(), "Oops", Toast.LENGTH_SHORT).show();
                        return true;
                    }
                });


        holder.createdTime.setText(currentMessage.getMsgTime());
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.message_user_name)
        TextView userName;
        @BindView(R.id.message_text)
        TextView messageText;
        @BindView(R.id.message_created)
        TextView createdTime;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/example/sasham/testproject/messages/MessagesListingAdapter.kt
package com.example.sasham.testproject.messages

import android.text.util.Linkify
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.example.sasham.testproject.Constants
import com.example.sasham.testproject.R
import com.example.sasham.testproject.model.Message
import com.example.sasham.testproject.util.StringUtil
import com.example.sasham.testproject.website.WebActivity
import com.squareup.picasso.Picasso
import de.hdodenhof.circleimageview.CircleImageView
import me.saket.bettermovementmethod.BetterLinkMovementMethod

class MessagesListingAdapter(private val messages: List<Message>) : androidx.recyclerview.widget.RecyclerView.Adapter<MessagesListingAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        val root = LayoutInflater.from(parent.context).inflate(R.layout.message_listing_item, parent, false)
        return ViewHolder(root)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {

        val context = holder.itemView.context
        val currentMessage = messages[position]

        holder.userName!!.text = currentMessage.userName

        holder.msgCount!!.text = if (StringUtil.isNotNullOrEmpty(currentMessage.msgCount))
            currentMessage.msgCount
        else
            "0"

        holder.messageText!!.text = StringUtil.stripHtml(currentMessage.msgText)
        holder.messageText!!.movementMethod = BetterLinkMovementMethod.getInstance()
        Linkify.addLinks(holder.messageText!!, Linkify.WEB_URLS)
        BetterLinkMovementMethod.linkify(Linkify.WEB_URLS, holder.messageText!!)
                .setOnLinkClickListener { textView, url ->
                    WebActivity.startActivity(url, context)
                    true
                }

        if (StringUtil.isNotNullOrEmpty(currentMessage.msgTime)) {
            holder.createdTime!!.text = StringUtil.getDateFromMillis(currentMessage.msgTime!!, Constants.TIME_PATTERN)
        }

        if (StringUtil.isNotNullOrEmpty(currentMessage.avatar)) {
            Picasso.get()
                    .load(StringUtil.getAvatarUrl(currentMessage.avatar!!))
                    .placeholder(R.mipmap.ic_avatar)
                    .into(holder.avatar)
        } else {
            holder.avatar!!.setImageResource(R.mipmap.ic_avatar)
        }
    }

    override fun getItemCount(): Int {
        return messages.size
    }

    inner class ViewHolder(val root: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(root) {


        var userName: TextView = root.findViewById(R.id.userName)

        var messageText: TextView = root.findViewById(R.id.msgText)

        var createdTime: TextView = root.findViewById(R.id.createdTime)

        var msgCount: TextView = root.findViewById(R.id.messagesCount)

        var avatar: CircleImageView = root.findViewById(R.id.avatar)

    }
}
